### PR TITLE
Deprecate Observable.return_type

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
`Observable.return_type` is a relic of an older PennyLane, and should be removed.

**Description of the Change:**
- raise a warning whenever `Observable.return_type` is accessed
- Observable reprs (like Tensor) will never try to add the return type to it
- change CircuitGraph to expect measurements instead of observables
- funky stuff with `tape._par_info[x]` -> I kept the info["op"] to be the observable (and not the surrounding measurement) because a lot of code expects it to be an operator (fair enough!). whenever code needed the return type, I instead get the `op_idx` from the info, then access the tape at that index to get the measurement
- `var_param_shift` now checks if it's replacing a trainable hamiltonian and fails before passing to `expval_param_shift`. This was a weird one: `expval_param_shift` would receive something like `qml.expval(H)`, but `H.return_type` would still be `Variance` because we never updated it when swapping measurements. The new way is to just fail right away.

**Benefits:**
- the property has been deprecated!
- Kept it as minimal as possible while avoiding using the property in PL source

**Possible Drawbacks:**
To keep things minimal, there are still some arguments named "observables" when they are actually supposed to be measurements. I can change that too if we want.

### CODE REVIEW

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -640,7 +632,13 @@ def statistics(
                 result = self._asarray(qml.math.squeeze(samples))
 
             elif isinstance(m, CountsMP):
-                result = self.sample(obs, shot_range=shot_range, bin_size=bin_size, counts=True)
+                result = self.sample(
+                    obs,
```
Comment (albi3ro) [inline pennylane/_qubit_device.py:636]: So obs could be either an observable or a measurement process?

Comment (timmysilv) [reply pennylane/_qubit_device.py:636]: no, it should still be an observable to match the other functions and not break the API. The observable eventually gets passed to `_samples_to_counts`, and on master this functions [depends on obs.return_type](https://github.com/PennyLaneAI/pennylane/blob/b7d6d6d184ef0b29256c13a4071f0a513c89b20c/pennylane/_qubit_device.py#L1446). I could maybe update the function to support both to avoid introducing breaking changes, but I'm worried that it'll cause other things to break

Comment (timmysilv) [reply pennylane/_qubit_device.py:636]: then again, only lightning overrides it. so I suppose we could make it support measurements or observables. either way... something has to change in lightning. I'd rather not support a Union-type because it feels sketchier (aka something only Python would let you do) but I can be convinced

Comment (albi3ro) [reply pennylane/_qubit_device.py:636]: So what if the measurement is `qml.sample()` without an observable?

Comment (timmysilv) [reply pennylane/_qubit_device.py:636]: oh crap good point. it already expects either one of them! updating, one sec.

File: pennylane/_qutrit_device.py [hunk 2]
```diff
@@ -357,6 +357,7 @@ def marginal_prob(self, prob, wires=None):
         prob = self._transpose(prob, np.argsort(np.argsort(device_wires)))
         return self._flatten(prob)
 
+    # pylint:disable=unused-argument,too-many-arguments
```
Comment (timmysilv) [inline pennylane/_qutrit_device.py:360]:
```suggestion
```

File: pennylane/devices/null_qubit.py [hunk 3]
```diff
@@ -291,6 +291,7 @@ def generate_samples(self):
         In the _qubit_device.py, the function calls for analytic_probability for its operations."""
         self.analytic_probability()
 
+    # pylint:disable=too-many-arguments
```
Comment (timmysilv) [inline pennylane/devices/null_qubit.py:294]:
```suggestion
```

File: pennylane/_qubit_device.py [hunk 4]
```diff
@@ -1465,6 +1454,7 @@ def circuit(x):
 
         return outcome_dicts if batched else outcome_dicts[0]
 
+    # pylint:disable=too-many-arguments
```
Comment (timmysilv) [inline pennylane/_qubit_device.py:1457]:
```suggestion
```

**Comment by timmysilv** [comment]
[sc-45858]

File: pennylane/_device.py [hunk 5]
```diff
@@ -421,7 +421,7 @@ def execute(self, queue, observables, parameters=None, **kwargs):
 
         Args:
             queue (Iterable[~.operation.Operation]): operations to execute on the device
-            observables (Iterable[~.operation.Observable]): observables to measure and return
+            observables (Iterable[~.measurements.MeasurementProcess]): observables to measure and return
```
Comment (trbromley) [inline pennylane/_device.py:424]: Can we avoid the breaking change in this release? E.g., support both `MeasurementProcesses` and `Observables`?

Comment (trbromley) [reply pennylane/_device.py:424]: This added complication will eventually be removed when the deprecation becomes a removal.

Comment (trbromley) [reply pennylane/_device.py:424]: Motivation: I think most of our plugins use `QubitDevice`, but I'd rather we avoid breaking `Device` for the sake of a deprecation.

File: pennylane/_qubit_device.py [hunk 6]
```diff
@@ -613,14 +612,7 @@ def statistics(
...
             # TODO: Remove this when all overriden measurements support the `MeasurementProcess` class
-            if m.obs is not None:
-                obs = m.obs
-                obs.return_type = m.return_type
-            elif m.mv is not None:
-                obs = m.mv
-                obs.return_type = m.return_type
-            else:
-                obs = m
+            obs = m.obs or m.mv or m
```
Comment (trbromley) [inline pennylane/_qubit_device.py:615]: What implications, if any, will this have on device developers?

File: pennylane/circuit_graph.py [hunk 7]
```diff
@@ -48,7 +49,7 @@ def _is_observable(x):
     Returns:
         bool: True iff x is an observable
     """
-    return getattr(x, "return_type", None) is not None
+    return isinstance(x, MeasurementProcess)
```
Comment (trbromley) [inline pennylane/circuit_graph.py:52]: Wouldn't this change behaviour so that a state-like measurement process is now classed as an observable?

File: pennylane/gradients/parameter_shift.py [hunk 8]
```diff
@@ -671,6 +671,16 @@ def var_param_shift(tape, argnum, shifts=None, gradient_recipes=None, f0=None, b
...
+        if obs.name == "Hamiltonian":
+            first_obs_idx = len(expval_tape.operations)
+            for t_idx in reversed(range(len(expval_tape.trainable_params))):
+                op, op_idx, _ = expval_tape.get_operation(t_idx)
+                if op_idx < first_obs_idx:
+                    break  # already seen all observables
+                if op is obs:
+                    raise ValueError(
+                        "Can only differentiate Hamiltonian coefficients for expectations, not variances"
+                    )
```
Comment (trbromley) [inline pennylane/gradients/parameter_shift.py:683]: Is this a new constraint? Or an already existing one that we didn't catch properly?

File: pennylane/operation.py [hunk 9]
```diff
@@ -1875,22 +1877,15 @@ def is_hermitian(self):
...
-        return (
-            f"{repr(self.return_type)}(wires={self.wires.tolist()})"
-            if self.return_type is qml.measurements.Probability
-            else f"{repr(self.return_type)}({temp})"
+    @property
+    def return_type(self):
+        """None or ObservableReturnTypes: Measurement type that this observable is called with."""
+        warnings.warn(
+            "`Observable.return_type` is deprecated. Instead, you should "
+            "inspect the type of the surrounding measurement process.",
```
Comment (trbromley) [inline pennylane/operation.py:1885]: Could we add an example?

File: pennylane/operation.py [hunk 10]
```diff
@@ -2109,17 +2103,7 @@ def __copy__(self):
...
-
-        if self.return_type is None:
-            return s
-
-        return (
-            f"{repr(self.return_type)}(wires={self.wires.tolist()})"
-            if self.return_type is qml.measurements.Probability
-            else f"{repr(self.return_type)}({s})"
-        )
+        return " @ ".join([repr(o) for o in self.obs])
```
Comment (trbromley) [inline pennylane/operation.py:2106]: What is the difference here?

**Review by trbromley — COMMENTED** [review]
Thanks @timmysilv, I have left a few questions.

File: pennylane/_device.py (hunk 5 above)
Comment (timmysilv) [reply pennylane/_device.py:424]: I'll give it a try. I think it's easier to just switch the PQ devices (I'm quite certain nothing else inherits from `Device` directly), and I don't feel so morally dirty doing it because the convention was not respected in the first place. Buuuut I'll see what I can do.

File: pennylane/_qubit_device.py (hunk 6 above)
Comment (timmysilv) [reply pennylane/_qubit_device.py:615]: I just used a shorthand to do the same as before, but without changing `obs.return_type` because we don't depend on it later anymore. I can be more verbose to check if it's `None`, but like... is `m.obs` ever `0`, `False`, or something else that's Falsey?

File: pennylane/circuit_graph.py (hunk 7 above)
Comment (timmysilv) [reply pennylane/circuit_graph.py:52]: this is an example of the first drawback I listed. I left the name `_is_observable`, but it's really just checking if it's a measurement. Before, if it was a tape, it would get the `qml.state()` as an "observable" when fetching `tape.observables`. Eg:
```pycon
>>> qs = qml.tape.QuantumScript([], [qml.state()])
>>> qs.observables, qs.observables[0].return_type
([state(wires=[])], state)
```
This function would have returned True in the past, and it still would.

File: pennylane/gradients/parameter_shift.py (hunk 8 above)
Comment (timmysilv) [reply pennylane/gradients/parameter_shift.py:683]: previously existing. see the point on `var_param_shift` in the PR description for why we didn't need to check it here before.

File: pennylane/operation.py (hunk 9 above)
Comment (timmysilv) [reply pennylane/operation.py:1885]: uhh. kinda but I'm not sure what it would look like? need a tape to explain why they might be different IMO
```pycon
>>> tape = qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0))])
>>> meas, obs = tape.measurements[0], tape.observables[0]
>>> print(meas, obs)
expval(PauliZ(wires=[0])) PauliZ(wires=[0])
>>> print(meas.return_type, obs.return_type)
pennylane/operation.py:1883: UserWarning: `Observable.return_type` is deprecated. Instead, you should inspect the type of the surrounding measurement process.
  warnings.warn(
ObservableReturnTypes.Expectation ObservableReturnTypes.Expectation
```

File: pennylane/operation.py (hunk 10 above)
Comment (timmysilv) [reply pennylane/operation.py:2106]: we include the return type on master (the thing being drawn is an `Observable`, not a `MeasurementProcess`):
```pycon
>>> qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))]).observables[0]
expval(PauliZ(wires=[0]) @ PauliZ(wires=[1]))
```

on this branch:
```pycon
>>> qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))]).observables[0]
PauliZ(wires=[0]) @ PauliZ(wires=[1])
```

this fixes the headache we mentioned while discussing this issue

File: pennylane/_device.py (hunk 5 above)
Comment (timmysilv) [reply pennylane/_device.py]: I changed it to be consistent with my change to `QubitDevice.sample`, and now it's fine with observables or measurements.

File: pennylane/operation.py (hunk 10 above)
Comment (trbromley) [reply pennylane/operation.py:2106]: Thanks, got it! This looks great.

File: pennylane/operation.py (hunk 9 above)
Comment (trbromley) [reply pennylane/operation.py:1885]: Thanks! Ok, let's leave it for now.

File: pennylane/circuit_graph.py (hunk 7 above)
Comment (trbromley) [reply pennylane/circuit_graph.py:52]: Thanks! Ok sounds good if this preserves the status quo.

**Review by mudit2812 — APPROVED** [review]
Looks great! Thanks Matt

File: pennylane/circuit_graph.py (hunk 7 above)
Comment (timmysilv) [reply pennylane/circuit_graph.py:52]: it does, but I do wanna be clear that there is a change [here](https://github.com/PennyLaneAI/pennylane/pull/4762/files#diff-fe39c5ae8018118cba5abadd13d54040cee7f520767ebf3459dec817ec8a2a91L978-R985). `QuantumScript.graph` used to construct a `CircuitGraph` using `self.observables` which we're moving away from, and this PR changes it to that property is constructed using `tape.measurements`. The logic is still exactly the same, but this file now uses the `MeasurementProcess` API instead of the `Observable` API.

In other words, if a tape's measurements are `[qml.expval(qml.PauliZ(0))]`:
- we used to construct the graph with a `PauliZ` Observable that has the (now deprecated) `return_type=expval` property set.
- we now construct the graph with an `ExpectationMP` instance that has the `self.obs` property set to the `PauliZ` Observable (the preferred API)

Comment (timmysilv) [reply pennylane/circuit_graph.py:52]: confirmed offline, will add an entry to breaking changes but otherwise this is fine.

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/_qubit_device.py]: it's updated. now I pass `m` to `self.sample`. in the past, if the measurement didn't _have_ an observable, it would have received the same object anyway, just under the name `obs` (which we defined by `obs = m` because `m.obs` was None). the private `_samples_to_counts` (the original source of trouble here) has been updated to expect measurement objects no matter what now (we know it will be a measurement because it's only called `if isinstance(m, CountsMP)`

**Review by albi3ro — APPROVED** [review]
Thanks for mucking around with this legacy spaghetti.

I've run this branch with lightning and qiskit, and it doesn't seem to cause any issues (though qiskit was broken for other reasons still).
